File: python-algo/leetcode/problem_82.py
"""82. Remove Duplicates from Sorted List II
https://leetcode.com/problems/remove-duplicates-from-sorted-list-ii/
"""
import collections
from typing import Optional

from common.list_node import ListNode


class Solution:
    def deleteDuplicates(self, head: Optional[ListNode]) -> Optional[ListNode]:
        if not head:
            return head
        dummy = ListNode(head.val - 1)
        pre = dummy
        cur = head
        while cur:
            if cur.next:
                if cur.val != pre.val and cur.val != cur.next.val:
                    pre.next = cur
                    cur = cur.next
                    pre = pre.next
                    pre.next = None
                else:
                    val = cur.val
                    while cur and cur.val == val:
                        cur = cur.next
            else:
                pre.next = cur
                cur = cur.next
        return dummy.next

File: java-algo/src/main/java/com/leetcode/Problem82.java
package com.leetcode;

import com.common.ListNode;

/**
 * 82. Remove Duplicates from Sorted List II
 * https://leetcode.com/problems/remove-duplicates-from-sorted-list-ii
 */
public class Problem82 {
    public ListNode deleteDuplicates(ListNode head) {
        if (head == null) {
            return null;
        }
        ListNode dummy = new ListNode(9999, head);
        ListNode a = dummy, b = head, c = head.next;
        while (c != null) {
            if (b.val != c.val) {
                a = b;
                b = c;
                c = c.next;
            } else {
                while (c != null && c.val == b.val) {
                    c = c.next;
                }
                b = c;
                a.next = b;
                if (c != null) c = c.next;
            }
        }
        return dummy.next;
    }
}
